src/mame/misc/sttechno.cpp

```diff
@@ -0,0 +1,608 @@
+}
+
+void sttechno_state::bank_w(uint16_t data)
+{
+	m_bank = data;
+}
+
+void sttechno_state::bank_write_enable_w(uint16_t data)
+{
+	// Only appears to be used for banks 3 and 4
+	// For bank 3 it'll write 1, for bank 4 it'll write 2
+	// The game typically enables write, writes data, then immediately clears the write flag
+	m_bank_write_enabled = data != 0;
+}
+
+void sttechno_state::data_w(offs_t offset, uint16_t data)
+{
+	if (m_bank >= 0 && m_bank <= 2) {
+		m_sound->write(offset + (0x100000 * m_bank), data);
+	} else if (m_bank == 3 && m_bank_write_enabled) {
+		m_flash[0]->write(offset, data);
+	} else if (m_bank == 4 && m_bank_write_enabled) {
+		m_flash[1]->write(offset, data);
+	}
+}
+
+uint16_t sttechno_state::data_r(offs_t offset)
+{
+	if (m_bank >= 0 && m_bank <= 2) {
+		return m_sound->read(offset + (0x100000 * m_bank));
+	} else if (m_bank == 3) {
+		return m_flash[0]->read(offset);
+	} else if (m_bank == 4) {
+		return m_flash[1]->read(offset);
+	}
+
+	return 0;
+}
+
+void sttechno_state::uart_data_w(uint16_t data)
+{
+	LOGMASKED(LOG_UART, "uart_data_w %04x\n", data);
+	transmit_register_setup(data & 0xff);
+}
+
+uint16_t sttechno_state::uart_data_r()
+{
+	uint16_t r = m_uart_rx;
+	m_uart_rx = 0;
+	m_maincpu->set_input_line(4, CLEAR_LINE);
+	m_uart_rx_busy = false;
+	return r;
+}
+
+void sttechno_state::tra_callback()
+{
+	m_rs232->write_txd(transmit_register_get_data_bit());
+}
+
+void sttechno_state::rcv_complete()
+{
+	receive_register_extract();
+	m_uart_rx = get_received_char();
+	m_uart_rx_busy = true;
+
+	m_maincpu->set_input_line(4, ASSERT_LINE);
+}
+
+uint16_t sttechno_state::uart_txrdy_r()
+{
+	// Polled before writing to uart_data_w
+	return !m_uart_rx_busy && is_transmit_register_empty() && !is_receive_register_full();
+}
+
+void sttechno_state::uart_config_w(uint16_t data)
+{
+	// 0x49 is written here
+	LOGMASKED(LOG_UART, "uart_config_w %04x\n", data);
+}
+
+void sttechno_state::ata_cs0_w(offs_t offset, uint16_t data, uint16_t mem_mask)
+{
+	if (!m_ata_enabled)
+		return;
+
+	m_ata->cs0_w(offset, data, mem_mask);
+}
+
+uint16_t sttechno_state::ata_cs0_r(offs_t offset, uint16_t mem_mask)
+{
+	if (!m_ata_enabled)
+		return 0;
+
+	return m_ata->cs0_r(offset, mem_mask);
+}
+
+void sttechno_state::ata_control_w(uint16_t data)
+{
+	if (data & 0x8000) {
+		// TODO: CD-ROM hardware reset
+		return;
+	}
+
+	m_ata_enabled = (data & 3) != 0;
+}
+
+uint16_t sttechno_state::video_unk_r()
+{
+	// This gets called at two points:
+	// - Every time before the OBJ RAM gets updated
+	// - Every time before setting FM_BYTEPROGRAM for the flash
+
+	if (m_sttga1_video_flash_write_enable)
+		return 0;
+
+	// HACK: Copy the OBJ data into the unused second half of the OBJ RAM for rendering
+	// This helps fix some small graphical glitches (sound test menu's text flicker, flickering during initial installation when programming the video flash)
+	std::copy_n(std::begin(m_sttga1_ram_obj), 0x2000, std::begin(m_sttga1_ram_obj) + 0x2000);
```

Comment:
Remember to check `!machine().side_effects_disabled()` for reads with side effects.

Reply:
Thanks, I always forget that. I found another that needs it too (uart_data_r).